src/delta.c

```diff
@@ -0,0 +1,397 @@
+ *
+ * This file is part of wolfBoot.
+ *
+ * wolfBoot is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * wolfBoot is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+#include <stdint.h>
+#include <string.h>
+#include "delta.h"
+
+
+#define ESC 0x7f
+
+struct __attribute__((packed)) block_hdr {
+    uint8_t esc;
+    uint8_t off[3];
+    uint8_t sz[2];
+};
+
+#define BLOCK_HDR_SIZE (sizeof (struct block_hdr))
+#define MAX_SRC_SIZE (1 << 24)
+
+#ifndef WOLFBOOT_SECTOR_SIZE
+#   define WOLFBOOT_SECTOR_SIZE 0x1000
+#endif
+
+int wb_patch_init(WB_PATCH_CTX *bm, uint8_t *src, uint32_t ssz, uint8_t *patch, uint32_t psz)
+
+{
+    if (!bm || ssz == 0 || psz == 0) {
+        return -1;
+    }
+    memset(bm, 0, sizeof(WB_PATCH_CTX));
+    bm->src_base = src;
+    bm->src_size = ssz;
+    bm->patch_base = patch;
+    bm->patch_size = psz;
+    return 0;
+}
+
+int wb_patch(WB_PATCH_CTX *ctx, uint8_t *dst, uint32_t len)
+{
+    struct block_hdr *hdr;
+    uint32_t dst_off = 0;
+    uint32_t src_off;
+    uint16_t sz;
+    uint32_t copy_sz;
+    if (!ctx)
+        return -1;
+    if (len < BLOCK_HDR_SIZE)
+        return -1;
+    if (ctx->p_off >= ctx->patch_size)
+        return 0;
+
+
+    while ( (ctx->p_off < ctx->patch_size) && (dst_off < len)) {
+        uint8_t *pp = (ctx->patch_base + ctx->p_off);
+        if (ctx->matching) {
+            /* Resume matching block from previous sector */
+            sz = ctx->blk_sz;
+            if (sz > len)
+                sz = len;
+            memcpy(dst + dst_off, ctx->src_base + ctx->blk_off, sz);
+            if (ctx->blk_sz > len) {
+                ctx->blk_sz -= len;
+                ctx->blk_off += len;
+            } else {
+                ctx->blk_off = 0;
+                ctx->blk_sz = 0;
+                ctx->matching = 0;
+            }
+            dst_off += sz;
+            continue;
+        }
+        if (*pp == ESC) {
+            if (*(pp + 1) == ESC) {
+                *(dst + dst_off) = ESC;
+                 /* Two bytes of the patch have been consumed to produce ESC */
+                ctx->p_off += 2;
+                dst_off++;
+                continue;
+            } else {
+                hdr = (struct block_hdr *)pp;
+                src_off = (hdr->off[0] << 16) + (hdr->off[1] << 8) +
+                    hdr->off[2];
+                sz = (hdr->sz[0] << 8) + hdr->sz[1];
```

Comment:
Same, consider big endian platform. Swapping might be different.

Reply:
Please note that the code is endian-agnostic here. The values are always stored in big endian format in `wb_diff` and always processed in the same byte order in `wb_patch`. 24-bit unsigned little-endian for the offset, 16-bit unsigned little endian for size. Values are never stored or retrieved directly using host type.